panther_hardware_interfaces/src/motors_controller.cpp

```diff
@@ -0,0 +1,246 @@
+: canopen_controller_(canopen_settings),
+  front_data_(drivetrain_settings),
+  rear_data_(drivetrain_settings),
+  roboteq_vel_cmd_converter_(drivetrain_settings),
+  pdo_feedback_timeout_(canopen_settings.pdo_feedback_timeout)
+{
+}
+
+void MotorsController::Initialize()
+{
+  try {
+    canopen_controller_.Initialize();
+  } catch (const std::runtime_error & e) {
+    throw e;
+  }
+}
+
+void MotorsController::Deinitialize() { canopen_controller_.Deinitialize(); }
+
+void MotorsController::Activate()
+{
+  // Activation procedure - it is necessary to first reset scripts, wait for a bit (1 second)
+  // and then send 0 commands for some time (also 1 second)
+
+  try {
+    canopen_controller_.GetFrontDriver()->ResetRoboteqScript();
+  } catch (const std::runtime_error & e) {
+    throw std::runtime_error(
+      "Front driver reset Roboteq script exception: " + std::string(e.what()));
+  }
+
+  try {
+    canopen_controller_.GetRearDriver()->ResetRoboteqScript();
+  } catch (const std::runtime_error & e) {
+    throw std::runtime_error(
+      "Rear driver reset Roboteq script exception: " + std::string(e.what()));
+  }
+
+  std::this_thread::sleep_for(std::chrono::milliseconds(1000));
+
+  try {
+    canopen_controller_.GetFrontDriver()->SendRoboteqCmdChannel1(0);
+    canopen_controller_.GetFrontDriver()->SendRoboteqCmdChannel2(0);
+  } catch (const std::runtime_error & e) {
+    throw std::runtime_error("Front driver send 0 command exception: " + std::string(e.what()));
+  }
+  try {
+    canopen_controller_.GetRearDriver()->SendRoboteqCmdChannel1(0);
+    canopen_controller_.GetRearDriver()->SendRoboteqCmdChannel2(0);
+  } catch (const std::runtime_error & e) {
+    throw std::runtime_error("Rear driver send 0 command exception: " + std::string(e.what()));
+  }
+  std::this_thread::sleep_for(std::chrono::milliseconds(1000));
+}
+
+void MotorsController::UpdateSystemFeedback()
+{
+  const RoboteqDriverFeedback front_driver_feedback =
+    canopen_controller_.GetFrontDriver()->ReadRoboteqDriverFeedback();
+  const RoboteqDriverFeedback rear_driver_feedback =
+    canopen_controller_.GetRearDriver()->ReadRoboteqDriverFeedback();
+
+  const timespec front_driver_ts = front_driver_feedback.timestamp;
+  const timespec rear_driver_ts = rear_driver_feedback.timestamp;
+  timespec current_time;
+  clock_gettime(CLOCK_MONOTONIC, &current_time);
+
+  const bool front_data_timed_out =
+    (lely::util::from_timespec(current_time) - lely::util::from_timespec(front_driver_ts) >
+     pdo_feedback_timeout_);
+  const bool rear_data_timed_out =
+    (lely::util::from_timespec(current_time) - lely::util::from_timespec(rear_driver_ts) >
+     pdo_feedback_timeout_);
+
+  const bool front_can_error = canopen_controller_.GetFrontDriver()->IsCanError();
+  const bool rear_can_error = canopen_controller_.GetRearDriver()->IsCanError();
+
+  // Channel 1 - right, Channel 2 - left
```

Comment:
```suggestion
  // Channel 1 - right motor, Channel 2 - left motor
```
Comment if fine as long as we assume this assignment won't ever change.